File: src/main/scala/com/benjaminrosenbaum/jovian/Geometry.scala
package com.benjaminrosenbaum.jovian
import scala.scalajs.js
import scala.scalajs.js.annotation._
import scala.annotation.meta.field
import scala.math._


//TODO decouple from spawning
object Plane { 
	val Width = 1500 
	val Height = 1800 
	val Buffer = 30
	val XSpan = new Span(0, Width)
	val YSpan = new Span(0, Height)

	def intoBottom(y : Double) = max(0, Buffer - (Height - y))/Buffer
	def intoRight(x : Double) = max(0, Buffer - (Width - x))/Buffer
	def intoLeft(x: Double) = max(0, Buffer - x)/Buffer
		
} 

abstract class Coords[T <: Coords[T]] (val x : Double, val y: Double) {
	//type T <: Coords

	val epsilon = 0.00001
	def copy(x : Double = x, y : Double = y) : T
	def plus[U <: Coords[U]](c: Coords[U]) = copy(c.x + x, c.y + y)
	def scaled(factor: Double) = copy(x * factor, y * factor)
	def equals(c: Coords[T]) = abs(x - c.x) + abs(y - c.y) < epsilon
	override def toString(): String = "(" + x + "," + y + ")"  	

	//this Coords object acts as the default object
	def averageAmong(ps : Seq[T]) : T = {
		val scale = if (ps.length == 0)  1.0 else (1.0 / ps.length)     
		val sum = ps.fold(this)((a, b) => a.plus(b))		 
		sum.scaled(scale)
	}
}

@JSExport
class Vector (
	@(JSExport @field) override val x : Double, 
	@(JSExport @field) override val y: Double) extends Coords[Vector](x,y) {
	//type T = Vector
	override def copy(x : Double = x, y : Double = y) = new Vector(x, y)
	def magnitude = sqrt(x * x + y * y);  //TODO memoize?
	def theta = atan2(y, x) - Pi/2
	def manhattanDist = abs(x) + abs(y)

	//vector between 0 and 1 in magnitude, for multiplying
	def normalized = if (equals(Vector.NULL)) Vector.NULL else scaled(1.0 / magnitude)
	def scaledTo(mag: Double) = normalized.scaled(mag);
	def capMagnitudeAt(mag: Double) = if (magnitude > mag) scaledTo(mag) else this  //NOTE possible performance optimization comparing w/square of mag

	//sum all the vectors, normalize to initial speed
	def changeDirectionBy(vs: Seq[Vector]) = vs.foldLeft(this)((v1,v2) => v1.plus(v2)).scaledTo(magnitude)
	
}

trait Position { 
	def pos : Point 
	//def to(p: Position) : Vector = pos.to(p)
}

@JSExport 
class Point (
	@(JSExport @field) override val x : Double, 
	@(JSExport @field) override val y: Double) extends Coords[Point](x,y) with Position {
	//type T = Point
	override def copy(x : Double = x, y : Double = y) = new Point(x, y)
	def to(p: Position) = new Vector(p.pos.x - x, p.pos.y - y)
	def distanceTo(p: Position) = if (equals(p.pos)) 0 else to(p).magnitude
	def isWithinManhattanDistanceOf[T <: Position](p : T, dist: Double) = to(p).manhattanDist <= dist
	def allWithinManhattanDistanceOf[T <: Position](ps : Seq[T], dist: Double) = ps.filter(isWithinManhattanDistanceOf(_, dist))
	override def pos = this

	def constrainToPlane : Point = 
		if (Plane.XSpan.contains(x) && Plane.YSpan.contains(y)) this else copy( (x + Plane.Width) % Plane.Width, min(max(0, y),Plane.Height))
}


//the line y = mx + b
abstract class Line (val m: Double, val b: Double) {
	def yFor(x: Double) = m * x + b
	def intersects(other : Line) = m == other.m;
	def intersectsAt(other: Line) : Option[Point] = 
		if (!intersects(other)) None
		else {
			val x = (other.b - b)/(m - other.m)	
			Some(new Point(x, yFor(x)))
		}
}

class Span(val lower: Double, val higher: Double) {
	assert(lower <= higher)
	def contains(d: Double): Boolean = d >= lower && d <= higher
	def contains(s: Span): Boolean = contains(s.lower) || contains(s.higher)
	def intersects(s: Span) =  contains(s) || s.contains(this)
	override def toString: String = lower + " -> " + higher
}


@JSExport
class Square (@(JSExport @field) val center: Point, @(JSExport @field) val side: Double ) {
	@JSExport val top    = center.y + side/2;
	@JSExport val left   = center.x - side/2;
	@JSExport val bottom = center.y - side/2;
	@JSExport val right  = center.x + side/2;
	val xSpan = new Span(left, right)
	val ySpan = new Span(bottom, top)
	def copy(center: Point = center, side: Double = side) = new Square(center, side)
	def contains(p: Point) : Boolean = p.x >= left && p.x <= right && p.y <= top && p.y >= bottom
	def intersects(s: Square) : Boolean = xSpan.intersects(s.xSpan) && ySpan.intersects(s.ySpan) 
	def manhattanDistTo(s: Square) : Double = center.to(s.center).manhattanDist
	def depthOfPenetration(s: Square): Double = {
		val d = manhattanDistTo(s)               //a ratio of how far "in" a colliding square has gotten
		if (d == 0) 0 else (side + s.side)/d     //touching at corners is 1, touching sides is 2, halfway in laterally is 4
	}
	override def toString: String = s"($left,$bottom)-($right,$top)"
}

object Point {
	val ORIGIN = new Point(0,0)
	def centerOf(ps : Seq[Point]) : Point = ORIGIN.averageAmong(ps)
	def apply(x: Double, y: Double): Point = new Point(x, y)
}

object Position {
	def centerOf(ps: Seq[Position]) : Point = Point.centerOf(ps.map(p => p.pos))
}

object Vector {
	val NULL = new Vector(0,0)
	def averageOf(vs : Seq[Vector]) : Vector = NULL.averageAmong(vs)
	def apply(x: Double, y: Double): Vector = new Vector(x, y)
}

File: geometry.js
//geometry.js

var Plane = { Width : 1500, Height : 1800, Buffer: 0};
Plane.extremity = function() { return new Point(Plane.Width, Plane.Height); }
Plane.nextPoint = function() {
	return new Point(Plane.Width * Math.random(), Plane.Height * Math.random() );
}
Plane.Limit = Plane.Width + Plane.Height;

function Coords (x, y) {
	this.x = x;
	this.y = y;
}
Coords.prototype.ctor = Coords;
Coords.prototype.magnitude = function() {
	return Math.sqrt(this.x* this.x + this.y*this.y);
}
Coords.prototype.scaled = function(factor) {
	return new this.ctor(this.x * factor, this.y * factor);
}
Coords.prototype.plus = function (coords) { //also works for adding a Vector to a Point
	return new this.ctor(this.x + coords.x, this.y + coords.y);
}
Coords.prototype.minus = function (coords) { //also works for adding a Vector to a Point
	return new this.ctor(this.x - coords.x, this.y - coords.y);
}
Coords.prototype.equals = function(coords) {
	var epsilon = 0.00001;
	return Math.abs(coords.x - this.x) + Math.abs(coords.y - this.y) < epsilon; //floating point
}
Coords.prototype.flipX = function() { return new this.ctor(-this.x, this.y); }
Coords.prototype.flipY = function() { return new this.ctor(this.x, -this.y); }
Coords.prototype.flip = function() { return new this.ctor(-this.x, -this.y); }

Coords.prototype.toString = function() {
	return "(" + this.x +"," + this.y +")";
}
Coords.averageOf = function(cs, initial) {  //zero length array will return initial; otherwise, return type is of objects in array
	var scale = (cs.length == 0) ? 1 : 1 / cs.length;
	return cs.reduce(function(acc, c){ return c.plus(acc); }, initial).scaled(scale);
}

//POINT
var Position = new Interface('Position', ['getPos']); 

function Point (x, y) {
	Coords.call(this, x, y);
	this.checkInterfaces();
}
//inherit 
Point.prototype = Object.create(Coords.prototype);
Point.prototype.ctor = Point;
implementsInterface(Point.prototype, Position);
Point.fromJSON = function(json) { return fromJSON(json, function(j) { return new Point(json.x, json.y);} ); } 
//duck interface Positioned
Point.prototype.getPos = function() { return this; } 
Point.getPoints = function(positions) { return positions.map(function(p) { return p.getPos(); }); }

// Rotates a point given cos(theta) and sin(theta). We don't pass in theta
// to avoid recomputation of cos and sin.
Point.prototype.rotated = function (cth, sth) {
	return new Point(cth * this.x - sth * this.y, sth * this.x + cth * this.y);
};
Point.prototype.to = function(position) {
	return new Vector(position.getPos().x-this.x, position.getPos().y-this.y);
}
Point.prototype.distanceTo = function(position) {
    if (this.equals(position.getPos())) return 0;
    return this.to(position.getPos()).magnitude();
};
//forced within vertical bounds, while horizontal bounds wrap. Not every point, just the center points of motiles.
Point.prototype.planified = function(plane) {
 	function wrapCoordinate(coord, bound) {
		if (coord < 0 ) return coord + bound;
		if (coord > bound) return coord - bound;
		return coord;
	}
	function limitCoordinate(coord, bound) {
		return Math.min(Math.max(coord, 0), bound);
	}
	return new Point(wrapCoordinate(this.x, plane.Width), limitCoordinate(this.y, plane.Height));
	//return new Point(limitCoordinate(this.x, plane.Width), limitCoordinate(this.y, plane.Height));
}
Point.prototype.allWithinManhattanDistanceOf = function(points, dist) { 
	var that = this; 
	return points.filter(function (p) {return that.to(p).manhattanDist() <= dist;}); 
}
Point.centerOf = function(positions) { return Coords.averageOf(positions.map(function(p){ return p.getPos(); }), Point.ORIGIN); }
Point.ORIGIN = new Point(0, 0);

/*Point.prototype.flipX = function() { return new Point(-this.x, this.y); }
Point.prototype.flipY = function() { return new Point(this.x, -this.y); }
Point.prototype.flip = function() { return new Point(-this.x, -this.y); }*/

//VECTOR  TODO extend Point?
function Vector (x, y) {
	Coords.call(this, x, y);
}
//inherit 
Vector.prototype = Object.create(Coords.prototype);
Vector.prototype.ctor = Vector;

Vector.prototype.theta = function () { return Math.atan2(this.y, this.x) - Math.PI/2; }
//vector between 0 and 1 in magnitude, for multiplying
Vector.prototype.normalized = function() {
    var mag = this.magnitude();
    return (mag == 0) ? Vector.NULL : this.scaled(1 / mag);
};
Vector.prototype.scaledTo = function(mag) { return this.normalized().scaled(mag);} //same direction, but scaled to given magnitude
Vector.prototype.capMagnitudeAt = function(mag) { return (this.magnitude() > mag) ? this.scaledTo(mag) : this;}
	  //TODO for performance, this could be redesigned to square the left instead of sqrt'ing the right
		  
Vector.prototype.manhattanDist = function() {	return Math.abs(this.x) + Math.abs(this.y);}
Vector.averageOf = function(vs) { return Coords.averageOf(vs, Vector.NULL); }
Vector.NULL = new Vector(0, 0);


function Rect(topLeft, bottomRight) {
	this.topLeft = topLeft;
	this.bottomRight = bottomRight;
	this.contains = function (p) {
		return p.x >= this.topLeft.x && p.x <= this.bottomRight.x 
		    && p.y >= this.topLeft.y && p.y <= this.bottomRight.y;
	}
	this.expand = function(s) {
		return new Rect(new Point(topLeft.x - s, topLeft.y - s), new Point(bottomRight.x + s, bottomRight.y + s));
	}
}



